src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,177 @@
+{
+    internal class RemoteJSDataStream : Stream
+    {
+        private readonly RemoteJSRuntime _runtime;
+        private readonly Guid _streamId;
+        private readonly long _totalLength;
+        private readonly CancellationToken _cancellationToken;
+        private readonly Stream _pipeReaderStream;
+        private readonly Pipe _pipe;
+        private long _bytesRead;
+
+        public static async Task<bool> SupplyData(RemoteJSRuntime runtime, string streamId, ReadOnlySequence<byte> chunk, string error)
+        {
+            if (!runtime.RemoteJSDataStreamInstances.TryGetValue(Guid.Parse(streamId), out var instance))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            await instance.SupplyData(chunk, error);
+            return true;
+        }
+
+        public static async Task<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = Guid.NewGuid();
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsDataReference, streamId, maximumIncomingBytes);
+            return remoteJSDataStream;
+        }
+
+        private RemoteJSDataStream(
+            RemoteJSRuntime runtime,
+            Guid streamId,
+            long totalLength,
+            long maxBufferSize,
+            CancellationToken cancellationToken)
+        {
+            _runtime = runtime;
+            _streamId = streamId;
+            _totalLength = totalLength;
+            _cancellationToken = cancellationToken;
+
+            _runtime.RemoteJSDataStreamInstances.Add(_streamId, this);
+
+            _pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize));
+            _pipeReaderStream = _pipe.Reader.AsStream();
+        }
+
+        // TODO: Surely this should be IAsyncEnumerable<ReadOnlySequence<byte>> so we can pass through the
+        // data without having to copy it into a temporary buffer. But trying this gives strange errors -
+        // sometimes the "chunk" variable below has a negative length, even though the logic in BlazorPackHubProtocolWorker
+        // never returns a corrupted item as far as I can tell.
+        private async Task SupplyData(ReadOnlySequence<byte> chunk, string error)
+        {
+            if (chunk.Length < 0)
+            {
+                throw new InvalidOperationException($"The incoming data chunk cannot be negative.");
```

Comment:
If we're changing this to a `byte[]` then there's no need for the negative-length check.